tests/rptest/tests/upgrade_test.py

```diff
@@ -0,0 +1,77 @@
+# Copyright 2022 Redpanda Data, Inc.
+#
+# Use of this software is governed by the Business Source License
+# included in the file licenses/BSL.md
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0
+
+import re
+
+from ducktape.utils.util import wait_until
+from rptest.tests.redpanda_test import RedpandaTest
+from rptest.services.cluster import cluster
+from rptest.services.redpanda import RESTART_LOG_ALLOW_LIST
+from rptest.services.redpanda_installer import RedpandaInstaller
+
+
+def wait_for_num_versions(redpanda, num_versions):
+    def get_unique_versions():
+        node = redpanda.nodes[0]
+        brokers_list = \
+            str(node.account.ssh_output(f"{redpanda.find_binary('rpk')} redpanda admin brokers list"))
+        redpanda.logger.debug(brokers_list)
+        version_re = re.compile("v\\d+\\.\\d+\\.\\d+")
+        return set(version_re.findall(brokers_list))
+
+    # NOTE: allow retries, as the version may not be available immediately
+    # following a restart.
+    wait_until(lambda: len(get_unique_versions()) == num_versions,
+               timeout_sec=30)
+    unique_versions = get_unique_versions()
+    assert len(unique_versions) == num_versions, unique_versions
+    return unique_versions
+
+
+class UpgradeFromSpecificVersion(RedpandaTest):
+    """
+    Basic test that upgrading software works as expected.
+    """
+    def __init__(self, test_context):
+        super(UpgradeFromSpecificVersion,
+              self).__init__(test_context=test_context, num_brokers=3)
+        self.installer = RedpandaInstaller(self.redpanda, self.test_context)
+
+    def setUp(self):
+        # NOTE: `rpk redpanda admin brokers list` requires versions v22.1.x and
+        # above.
+        self.installer.start()
+        self.installer.install(self.redpanda.nodes, (22, 1, 3))
```

Comment:
If we create the installer within RedpandaService then we can also move these setup steps to another location such as RedpandaTest. The consequence of this change is that future test writers do not have to edit `setUp` for individual tests.

Reply:
I suspect a user would have to overwrite `setUp` anyway to specify the version. For numeric versions this is straightforward and could probably happen in the constructor. When using `installer.highest_from_prior_feature_version()` as the initial install version, that isn't the case (at least as implemented, requires an already-started installer).

Reply:
Fair point about putting `start()` in `setUp()` I'll consider it, though I was initially hesitant to push this into RedpandaTest since a vast minority of RedpandaTests would actually use it. That said, it looks like we have some other similar harnesses there.

Reply:
Actually it's a bit more nuanced than that. `RedpandaInstaller::start()` must be called before `install()`, and `install()` must be called before `RedpandaService::start()`.

I could begin tracking a desired version in the `RedpandaTest` constructor so that when it comes time to `setUp`, it's clear what versions need to be installed, but I don't love the idea of smearing version tracking across test abstractions, and this falls short when considering `installer.highest_from_prior_feature_version()` style versions.

I'm going to keep this as is for now. LMK If you have suggestions or feel strongly about it. I'm hoping this is a simple enough API as is and that there are clear error messages when users misuse it.

Reply:
> Actually it's a bit more nuanced than that. `RedpandaInstaller::start()` must be called before `install()`, and `install()` must be called before `RedpandaService::start()`.

This makes sense, I do not feel strongly about the changes.